### PR TITLE
WIP: [model] refactor: [1/N] Move model related code into DataParallelPPOActor

### PR DESCRIPTION
**Pull request by vermouth1992**
### Checklist Before Starting

- [x] Search for similar PR(s).

### What does this PR do?

We are in the process of refactor to implement Model as an Engine so that the training part of the code will become more modular and reusable.

To implement Model as an Engine, we propose the following high level steps to enable smooth migration without breaking APIs and make code review much easier.
1. Move model related code into DataParallelPPOActor
2. Abstract DataParallelPPOActor into DataParallelModelEngine and a set of data preprocess/postprocess/loss function specific to PPOActor
3. Implement DataParallelPPOCritic using DataParallelModelEngine and critic specific data preprocess/postprocess/loss function
4. Implement RewardModelWorker using DataParallelModelEngine and ORM specific data preprocess/postprocess/loss function
5. Replace the OmegaConfig with explicit dataclass config

This is part to implement the first step:
- move model construct inside DataParallelPPOActor
- move FSDPCheckpointManager inside DataParallelPPOActor
- move lora config into DataParallelPPOActor
- move load/offload into DataParallelPPOActor

In the next PR, ulysses data processing will be moved into the DataParallelPPOActor

### High-Level Design

> Demonstrate the high-level design if this PR is complex.

### Specific Changes

> List the specific changes.

### API

> Demonstrate how the API changes if any.

### Usage Example

> Provide usage example(s) for easier usage.

```python
# Add code snippet or script demonstrating how to use this 
```

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluatuion results, etc.

### Additional Info.

- **Issue Number**: Fixes issue # or discussion # if any.
- **Training**: [Note which backend this PR will affect: FSDP, Megatron, both, or none]
- **Inference**: [Note which backend this PR will affect: vLLM, SGLang, both, or none]

### Checklist Before Submitting

- [ ] Read the [Contribute Guide](https://github.com/volcengine/verl?tab=readme-ov-file#contribution-guide).
- [ ] Apply [pre-commit checks](https://github.com/volcengine/verl?tab=readme-ov-file#code-linting-and-formatting).
- [ ] Add `[BREAKING]` to the PR title if it breaks any API.
- [ ] Update the documentation about your changes in the [docs](https://github.com/volcengine/verl/tree/main/docs).
- [ ] New CI unit test(s) are added to cover the code path.
- [ ] Rely on existing unit tests on CI that covers the code path.

### CODE REVIEW

File: verl/workers/actor/dp_actor.py [hunk 1]
```diff
@@ -74,6 +124,318 @@ def __init__(self, config, actor_module: nn.Module, actor_optimizer: torch.optim
...
+        # so that DataParallelActor can be instantiated solely base on config. We keep compatible with old approach
+        # that constructing model and optimizer, when pass them as arguments. We will remove the compatibility in v0.6
+        self.rank = torch.distributed.get_rank()
+        self.inference_only = inference_only
+        self._lora_rank = self.config.model.get("lora_rank", 0)
+        self._is_lora = self._lora_rank > 0
+        # build device mesh for FSDP
+        world_size = torch.distributed.get_world_size()
+        # TODO(sgm): support FSDP hybrid shard for larger model
+        self.device_mesh = create_device_mesh(world_size=world_size, fsdp_size=self.config.fsdp_config.fsdp_size)
```
Comment (eric-haibin-lin) [inline verl/workers/actor/dp_actor.py:137]: the device mesh was still in https://github.com/volcengine/verl/blob/0eae03ae2bb369012c230edcb44de244e520e12b/verl/workers/fsdp_workers.py#L90

**Review by eric-haibin-lin — COMMENTED** [review]
@ZihengJiang fyi..

File: verl/workers/actor/dp_actor.py [hunk 2]
```diff
@@ -50,8 +79,29 @@
...
+    elif device_mesh.ndim == 2:
+        sharding_strategy = ShardingStrategy.HYBRID_SHARD
+    else:
+        raise NotImplementedError(f"Get device mesh ndim={device_mesh.ndim}, but only support 1 or 2")
+    return sharding_strategy
+
+
 class DataParallelPPOActor(BasePPOActor):
-    def __init__(self, config, actor_module: nn.Module, actor_optimizer: torch.optim.Optimizer = None):
+    def __init__(self, config, actor_module: nn.Module = None, actor_optimizer: torch.optim.Optimizer = None, inference_only=False, tokenizer=None, processor=None):
```
Comment (eric-haibin-lin) [inline verl/workers/actor/dp_actor.py:104]: is it possible to create a unit test for dp actor now?

Comment (vermouth1992) [reply verl/workers/actor/dp_actor.py:104]: Yes. Let's on hold this PR and discuss the details until we reach a clear path of transition

Comment (dataproblems) [inline verl/workers/actor/dp_actor.py:104]: Can you help me understand - in what cases would we have a need for initializing a DataParallelPPOActor with no nn.Module? Also is there a reason to not use Optional[nn.Module] instead?

File: verl/workers/actor/dp_actor.py [hunk 3]
```diff
@@ -74,6 +124,318 @@ def __init__(self, config, actor_module: nn.Module, actor_optimizer: torch.optim
         )
         self.device_name = get_device_name()
 
+        # Note(vermouth1992): we move the device mesh constructor into the DataParallelActor
+        # so that DataParallelActor can be instantiated solely base on config. We keep compatible with old approach
+        # that constructing model and optimizer, when pass them as arguments. We will remove the compatibility in v0.6
+        self.rank = torch.distributed.get_rank()
```
Comment (dataproblems) [inline verl/workers/actor/dp_actor.py:130]: Let's create strategies instead? 

For e.g. 

```
class OptimizerStrategy(ABC):
    def __init__(
        self, *, role: str, sharded_model: nn.Module, optim_config: Optional[DictConfig] = None
    ):
        self._sharded_model = sharded_model
        self.role = role
        self.optim_config = optim_config
        self.optimizer = None

    @property
    def optimizer(self):
        pass

    def zero_grad(self):
        if self.optimizer is not None:
            self.optimizer.zero_grad(set_to_none=True)
        else:
            raise RuntimeError("Optimizer not initialized.")

```

Then use it in the constructor to compose the actor instead of relying on the actor constructor to do everything? 


Similar logic for parallelism strategy to deal with fsdp / megatron.

**Review by dataproblems — CHANGES_REQUESTED** [review]

File: verl/workers/actor/dp_actor.py (hunk 2 above)
Comment (vermouth1992) [reply verl/workers/actor/dp_actor.py:104]: This is for backward compatible. The original design of DataParallelPPOActor requires users to init the FSDP model outside DataParallelPPOActor. Now we move the init inside DataParallelPPOActor if actor_module is None.

File: verl/workers/actor/dp_actor.py (hunk 3 above)
Comment (vermouth1992) [reply verl/workers/actor/dp_actor.py:130]: Nice suggestion. Let's do this one PR after another PR. For refactor, each PR should only modify a small portion of code to make review and testing easy.

File: verl/workers/actor/dp_actor.py (hunk 2 above)
Comment (dataproblems) [reply verl/workers/actor/dp_actor.py:104]: I see - perhaps in the future we can try to make it explicit? I'm still not sure I understand the use case where the nn.Module would be None ( in that case, what would the actor do? )
